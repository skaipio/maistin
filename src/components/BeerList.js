import React, { Component } from 'react';
import { ListGroup, ListGroupItem } from 'reactstrap';

export default class BeerList extends Component {
  constructor({ beers }) {
    super();
    this.state = {
      beers: JSON.parse(JSON.stringify(beers)), // deep clone
    };
    this.toggleFavorite = this.toggleFavorite.bind(this);
  }

  toggleFavorite(beerIndex) {
    const beers = JSON.parse(JSON.stringify(this.state.beers));
    beers[beerIndex].favorite = !beers[beerIndex].favorite;
    this.setState({
      beers: beers,
    });
  }

  render() {
    return (
      <ListGroup>
        { this.state.beers.map((beer, index) =>
          <ListGroupItem key={beer.id} onClick={() => this.toggleFavorite(index)}>{ beer.name }{ beer.favorite ? ' <3' : ''}</ListGroupItem>,
        ) }
      </ListGroup>
    );
  }
}
