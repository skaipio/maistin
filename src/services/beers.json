[{
  "id": 1,
  "name": "Ginger Ale",
  "favorite": false
},
{
  "id": 2,
  "name": "Nutty Stout",
  "favorite": true
}]