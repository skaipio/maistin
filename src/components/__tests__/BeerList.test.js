import React from 'react';
import { shallow } from 'enzyme';
import BeerList from '../BeerList';
import beers from './beers.json';

const keyEquals = (strKey, intKey) => parseInt(strKey, 10) === intKey;

describe('<BeerList />', () => {
  let mockBeers;

  beforeEach(() => {
    mockBeers = JSON.parse(JSON.stringify(beers));
  });

  it('renders the beers passed to it', () => {
    const wrapper = shallow(<BeerList beers={mockBeers} />);
    beers.forEach((beer) => {
      const beerItem = wrapper.find('li').filterWhere(li => keyEquals(li.key(), beer.id));
      expect(beerItem.length).toBe(1);
    });
  });

  it('displays no heart icon for a beer that is not favorited', () => {
    const wrapper = shallow(<BeerList beers={mockBeers} />);
    beers.filter(beer => !beer.favorite)
      .forEach((beer) => {
        const beerItem = wrapper.find('li').filterWhere(li => keyEquals(li.key(), beer.id)).first();
        expect(beerItem.text()).not.toContain('<3');
      });
  });

  it('displays a heart icon for a favorited beer', () => {
    const wrapper = shallow(<BeerList beers={mockBeers} />);
    beers.filter(beer => beer.favorite)
      .forEach((beer) => {
        const beerItem = wrapper.find('li').filterWhere(li => keyEquals(li.key(), beer.id)).first();
        expect(beerItem.text()).toContain('<3');
      });
  });

  it('displays a heart icon after clicking on a beer without a heart icon', () => {
    const wrapper = shallow(<BeerList beers={mockBeers} />);
    const beerNotFavorited = 1;
    let beerItem = wrapper.find('li').filterWhere(li => keyEquals(li.key(), beerNotFavorited)).first();
    beerItem.simulate('click');
    beerItem = wrapper.find('li').filterWhere(li => keyEquals(li.key(), beerNotFavorited)).first();
    expect(beerItem.text()).toContain('<3');
  });

  it('displays no heart icon after clicking on a beer with a heart icon', () => {
    const wrapper = shallow(<BeerList beers={mockBeers} />);
    const beerFavorited = 2;
    let beerItem = wrapper.find('li').filterWhere(li => keyEquals(li.key(), beerFavorited)).first();
    beerItem.simulate('click');
    beerItem = wrapper.find('li').filterWhere(li => keyEquals(li.key(), beerFavorited)).first();
    expect(beerItem.text()).not.toContain('<3');
  });
});
