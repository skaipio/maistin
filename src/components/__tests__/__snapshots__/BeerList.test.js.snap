// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BeerList /> snapshot of beer list matches 1`] = `
<ul
  className="list-group"
>
  <li
    className="list-group-item"
    onClick={[Function]}
  >
    Lovely Lager
    
  </li>
  <li
    className="list-group-item"
    onClick={[Function]}
  >
    Sound Stout
     &lt;3
  </li>
  <li
    className="list-group-item"
    onClick={[Function]}
  >
    Pounding Porter
    
  </li>
  <li
    className="list-group-item"
    onClick={[Function]}
  >
    Awesome Ale
     &lt;3
  </li>
</ul>
`;
